### DISCUSSION TITLE
Web Terminal with remote clusters

### DISCUSSION THREAD

**Discussion by gaeljw** [argoproj/argo-cd#15813] [Q&A] [answered]
Hello folks,

I'm evaluating the "web terminal" feature to ease the lives of our developers and let them exec into pods.

Currently, we have 1 ArgoCD installation running in cluster A that manages both cluster A and cluster B.

I'd like to somehow enforce that people will be able to use web terminal feature only for pods running in cluster B but not for cluster A.

Reading the [documentation](https://argo-cd.readthedocs.io/en/stable/operator-manual/web_based_terminal/), I see there's a `Role`/`ClusterRole` to be patched. It makes sense for exec-ing into pods of the same cluster as ArgoCD is running. But this cannot work for exec-ing into pods of another cluster, right?

I see that there's a `ClusterRole` `argocd-manager-role` in my cluster B. Sounds like I'd need to modify this one, right?

Then, the next question would be: where does this `argocd-manager-role` comes from? I don't remember but I guess it was setup when we used `argocd cluster add` for adding cluster B? The rights it currently doesn't seem right to me (too much rights), is there any documentation at what minimum rights it should have?

Thanks,

**Comment by gaeljw** [comment, marked as answer]
Answering myself: indeed if the `argocd-manager-role` `ClusterRole` in the remote cluster has proper privileges/rights, it works fine.

**Comment by gaeljw** [comment]
Opening another discussion for my second question: https://github.com/argoproj/argo-cd/discussions/15821

**Comment by zMynxx** [comment]
> Answering myself: indeed if the `argocd-manager-role` `ClusterRole` in the remote cluster has proper privileges/rights, it works fine.

Did you manage to patch that role using kustomize by any chance?

**Reply by gaeljw** [reply]
Well, I think it is not created by the ArgoCD install. It's created manually by yourself when adding the new remote cluster somehow.

Thus, I believe you don't have any Kustomize patch to apply. You could directly create the expected role.

**Reply by zMynxx** [reply]
My clusters are added declaratively using secrets, therefore I'm not really sure about how the create this ClusterRole. 
Either way I prefer doing so by patching the ArgoCD on initialization as part of the bootstrap.
I'll explore and update 😄

**Reply by gaeljw** [reply]
The thing is ArgoCD won't be able to create the role in the remote cluster if it doesn't have access to the remote cluster yet. At some point you'll have something to do "outside of Argo" I think. But it can surely be automated in your cluster installation process.

**Reply by zMynxx** [reply]
Well, I deployed everything again, no clusterrole / role had been created in the remote clusters, therefor no permissions are granted.
Could you walk me through what is needed? How can I allow Argo's ClusterRole to have permissions in remote clusters?

**Reply by gaeljw** [reply]
I used the CLI `argocd cluster add` but there may be other ways.

**Reply by zMynxx** [reply]
Follow up: I've doubled up and used `argocd cluster add` to init the sa-clusterrole and binding process on the remote cluster.
from observing the manifests, its permissions are sufficient enough.
There are the manifest generated by argocd on the remote cluster:
argocd-manager ServiceAccount:
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  creationTimestamp: "2024-04-15T11:41:33Z"
  name: argocd-manager
  namespace: kube-system
  resourceVersion: "13229"
  uid: 1e1b52ed-d5ba-4428-a2e7-ba6604685827
secrets:
- name: argocd-manager-token-h4cwb
```

argocd-manager-role ClusterRole:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  creationTimestamp: "2024-04-15T11:41:33Z"
  name: argocd-manager-role
  resourceVersion: "13218"
  uid: 22a301b1-420b-444b-bad8-615221fb669c
rules:
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - '*'
- nonResourceURLs:
  - '*'
  verbs:
  - '*'
```

and the argocd-manager-role-binding ClusterRoleBinding:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  creationTimestamp: "2024-04-15T11:41:33Z"
  name: argocd-manager-role-binding
  resourceVersion: "13219"
  uid: dca300af-ba26-49ea-94f7-62e71e67c50b
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: argocd-manager-role
subjects:
- kind: ServiceAccount
  name: argocd-manager
  namespace: kube-system
```

However, still unable to exec to the remote cluster workloads. Using custom accounts as well as the admin account.
I can however exec into current cluster workloads, so that on point for sure. Any guesses where I should look?
I also tried explicitly adding 
```yaml
- apiGroups:
  - ''
  resources:
  - 'pods/exec'
  verbs:
  - 'create'
 ```
to the clusterrole, no luck there..

**Reply by gaeljw** [reply]
When you say you're unable to "exec". Do you see the "terminal" tab in ArgoCD UI and it's black or you don't see the tab at all?

Role look's good. I have a similar setup and it works fine.

**Reply by zMynxx** [reply]
Yeah, here's some more info:
Screenshot:
<img width="938" alt="Screenshot 2024-04-15 at 16 38 55" src="https://github.com/argoproj/argo-cd/assets/124557045/b055d49d-bc70-464f-9296-ddc08530c64a">

Logs from argocd-server:
```
time="2024-04-15T12:57:44Z" level=info msg="terminal session starting" appNamespace=argocd application=demo-app-production container=webserver-simple namespace=demo-app podName=prod-simple-deployment-5df4f59c7b-gq95p project=production userName=admin
2024/04/15 12:57:44 http: response.WriteHeader on hijacked connection from github.com/argoproj/argo-cd/v2/server/application.(*terminalHandler).ServeHTTP (terminal.go:253)
2024/04/15 12:57:44 http: response.Write on hijacked connection from fmt.Fprintln (print.go:305)
time="2024-04-15T12:57:44Z" level=error msg="read message err: read tcp 127.0.0.1:8080->127.0.0.1:49700: use of closed network connection"
E0415 12:57:44.455389      10 v2.go:104] EOF
time="2024-04-15T12:57:44Z" level=error msg="read message err: read tcp 127.0.0.1:8080->127.0.0.1:49700: use of closed network connection"
E0415 12:57:44.455416      10 v2.go:104] use of closed network connection
```

My guess it's a networking issue, trying to resolve the wrong endpoint. I'll keep on investigating this way.

**Reply by zzyco** [reply]
Hey @zMynxx any resolutions? I am facing with the same issue
